CorsixTH/Lua/objects/helicopter.lua

```diff
@@ -82,17 +82,25 @@ function Helicopter:spawnPatient()
   local patient = self.world:newEntity("Patient", 2)
   patient:setDisease(hospital.emergency.disease)
   patient.diagnosis_progress = 1
+  patient.is_emergency = self.spawned_patients
   patient:setDiagnosed()
   patient:setMood("emergency", "activate")
-  patient.is_emergency = self.spawned_patients
-  hospital.emergency_patients[#hospital.emergency_patients + 1] = patient
+  hospital.emergency_patients[self.spawned_patients] = patient
   local x, y = hospital:getHeliportSpawnPosition()
   patient:setNextAction(SpawnAction("spawn", {x = x, y = y}):setOffset({y = 1}))
   patient:setHospital(hospital)
   -- TODO: If new combined diseases are added this will not work correctly anymore.
   patient.cure_rooms_visited = #patient.disease.treatment_rooms - 1
   local no_of_rooms = #patient.disease.treatment_rooms
-  patient:queueAction(SeekRoomAction(patient.disease.treatment_rooms[no_of_rooms]))
+  -- a spawning emergency patient might rather leave than pay
+  if not patient:agreesToPay(patient.disease.id) then
```

Comment:
agreesToPay here is based on the current price, which - if not the first patient (as set in your condition below) might not match.

Better to set the emergency treatment_price before this check and perhaps pass it into agreesToPay (which would then have to filter through priceDistortion)

Reply:
TH does it that way, sets the price then calls the equivalent of agreesToPay based on the current price, not the emergency price. In priceDistortion, there are 3 factors, happiness is equal to 1, if the there is no effectiveness for the treatment its '100%' and takes it away so then just becomes a weighting on the reputation - proxy for price anyway. Effectiveness is proxied into reputation as well as more effective less chance of death, higher rep.

TH agreesToPay equivalent is a little different as its only taking into account the 'fair' cost (price * rep / 500) then checks to see if the 'asking' price is less than random('fair' cost, 1.5 * 'fair' cost) and if it is agree. The other mechanic in TH which we don't have, is that when insurance pays 'fair' cost is inflated by a percentage which is unique per insurance company. And insurance companies pay out at different times as well. I haven't checked to see if the inflated prices may actually be agreed on or just added to the insurance payouts, I assume they should.

The other part of agreesToPay in TH (not on emergnecy), is when the patients agree to pay, they are also agreeing on a price that may change in favour or not, rather than the price agreed. TH saves the agreed to price on the patient and thats what is paid. Thus after that decision is made you are free to change the price in Corsix with the only ramification that you alter the perception for new patients, but you would have to be keen to do this on a persistent basis, maybe if your room crashed and you had patients waiting again you can do this trick to recover some build cost nearly immediately.